tooling/image-updater/internal/updater/updater.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2025 Microsoft Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package updater
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/go-logr/logr"
+
+	"github.com/Azure/ARO-HCP/tooling/image-updater/internal/clients"
+	"github.com/Azure/ARO-HCP/tooling/image-updater/internal/config"
+	"github.com/Azure/ARO-HCP/tooling/image-updater/internal/yaml"
+)
+
+// ImageUpdate represents a successful image update
+type ImageUpdate struct {
+	Name      string
+	NewDigest string
+}
+
+// Updater contains all pre-created resources needed for execution
+type Updater struct {
+	Config          *config.Config
+	DryRun          bool
+	RegistryClients map[string]clients.RegistryClient
+	YAMLEditors     map[string]*yaml.Editor
+	Updates         []ImageUpdate
+}
+
+// New creates a new Updater with all necessary resources pre-initialized
+func New(cfg *config.Config, dryRun bool, registryClients map[string]clients.RegistryClient, yamlEditors map[string]*yaml.Editor) *Updater {
+	return &Updater{
+		Config:          cfg,
+		DryRun:          dryRun,
+		RegistryClients: registryClients,
+		YAMLEditors:     yamlEditors,
+		Updates:         []ImageUpdate{},
+	}
+}
+
+// UpdateImages processes all images in the configuration using pre-created resources
+func (u *Updater) UpdateImages(ctx context.Context) error {
+	for name, imageConfig := range u.Config.Images {
+		digest, err := u.fetchLatestDigest(imageConfig.Source)
+		if err != nil {
+			return fmt.Errorf("failed to fetch latest digest for %s: %w", name, err)
+		}
+
+		for _, target := range imageConfig.Targets {
+			if err := u.updateImage(ctx, name, digest, target); err != nil {
+				return fmt.Errorf("failed to update image %s: %w", name, err)
+			}
+		}
+	}
+
+	if !u.DryRun && len(u.Updates) > 0 {
+		commitMsg := u.GenerateCommitMessage()
+		if commitMsg != "" {
+			fmt.Println(commitMsg)
+		}
+	}
+
+	return nil
+}
+
+// fetchLatestDigest retrieves the latest digest from the appropriate registry
+func (u *Updater) fetchLatestDigest(source config.Source) (string, error) {
+	registry, repository, err := source.ParseImageReference()
+	if err != nil {
+		return "", fmt.Errorf("failed to parse registry from image reference: %w", err)
+	}
+
+	client, exists := u.RegistryClients[registry]
+	if !exists {
+		return "", fmt.Errorf("no registry client available for %s", registry)
+	}
+
+	return client.GetLatestDigest(repository, source.TagPattern)
```

Comment:
This needs to ensure that the digest is for an image with the correct target architecture (or the multi-arch one)